### ISSUE TITLE
New-Guid Improvement

### ISSUE THREAD

**Issue by gw-bcharboneau** [PowerShell/PowerShell#19950]
### Summary of the new feature / enhancement

This is to expand on current functionality of Microsoft.PowerShell.Utility\New-Guid. I would like to see the EmptyValue and ValueOnly Options added to New-Guid Cmdlet.

- EmptyValue would provide a GUID Object of all Zeros
- ValueOnly would take a string GUID passed in on the commandline and generate it as a GUID Object

What this would solve is currently the New-guid cmdlet only generates a new Guid. But it could also generate an Empty Guid as well as format a Guid string as a Guid object.

I am not strongly familiar with C Languages, so I cannot provide an example on how to implement this in the cs file for the New-Guid CmdLet. However, I provide my example in powershell function form so that someone may be able to help improve it.

Currently Users have to use `[GUID]::Empty` for Empty Guids, `[GUID]($Value)` for converting guid string to guid object and `[GUID]::NewGuid()` or the existing `New-Guid` cmdlet to generate a new Guid. adding the 2 additional parameters to New-Guid would allow users to use it for all guid generation needs. This provides some standardization in Guid Generation.

I do however concede, that for the ValueOnly Option, a ConvertTo-Guid cmdlet could also be made. but for EmptyValue New-Guid seems appropriate.

### Proposed technical implementation details (optional)

```
function New-Guid() {
    [CmdletBinding(DefaultParameterSetName = 'New')]
    param (
        [Parameter(ParameterSetName = 'Empty')]
        [switch]
        $EmptyValue,
        [Parameter(ParameterSetName = 'Value')]
        [string]
        $ValueOnly
    )
    if($EmptyValue) {
        $guid = [GUID]::Empty
    } elseif ($ValueOnly) {
        $guid = [GUID]($Value)
    } else {
        $guid = [GUID]::NewGuid()
    }
    return $guid
}
```

**Comment by iRon7** [comment]
@gw-bcharboneau,

You didn't write a:
> **Summary of the new feature / enhancement**
> A clear and concise description of what the problem is that the new feature would solve. Try formulating it in user story style (if applicable).

Which would probably make it hard for the PowerShell team to value your feature / enhancement request.
But I personally do see some value in the `-Value` parameter where I would actually call it [`-ValueOnly`](https://learn.microsoft.com/powershell/module/microsoft.powershell.utility/get-variable#-valueonly) for consistency with other cmdlets.

> i do not have the existing code for the New-Guid Cmdlet

**For your background:**
You might refer to back to the original `New-Guid` cmdlet by using its [qualified name](https://learn.microsoft.com/powershell/module/microsoft.powershell.core/about/about_command_precedence#using-module-qualified-names):

```PowerShell
    } else {
        $guid = Microsoft.PowerShell.Utility\New-Guid
    }
```
(You might even consider to build [proxy command](https://devblogs.microsoft.com/scripting/proxy-functions-spice-up-your-powershell-core-cmdlets/)s to create prototypes which doesn't make much sense for this specific `New-Guid` example as there is no input, but might help you for more complex cmdlets)

**Comment by gw-bcharboneau** [comment]
> 

I have updated the original summary with more information

**Comment by JamesWTruher** [comment]
i really don't care for the name `ValueOnly` for that string parameter. It seems like it should be `FromString` or something along that line. FWIW, we already have a `ValueOnly` parameter on `Get-Variable` (it's a switch) and it returns the value of the variable rather than the PSVariable object

**Comment by TobiasPSP** [comment]
The cmdlet WG reviewed this and feels it is a useful enhancement. For consistency reasons, we suggest to name the parameter *-AsString* rather than *-ValueOnly*.

**Comment by gw-bcharboneau** [comment]
I am fine with whatever Alterations/Suggestions the team agrees upon.  glad to hear it will be added, eventually.

**Comment by ArmaanMcleod** [comment]
I personally don't find using `-AsString` consistent at all, given that parameter is a `switch` for other utility cmdlets like `Group-Object` and `Get-Unique`, and here it seems we are intending to use it for string input.

Any reason why we can't just rely on `New-Guid -InputObject` which already takes a string and converts to Guid? Or am I missing something here?

**Comment by iSazonov** [comment]
> Any reason why we can't just rely on `New-Guid -InputObject` which already takes a string and converts to Guid? Or am I missing something here?

The cmdlet hasn't `InputObject` parameter. The parameter is mainly for pipeline scenario. In the case it is better to have `ConvertTo-Guid` cmdlet.

**Comment by ArmaanMcleod** [comment]
> > Any reason why we can't just rely on `New-Guid -InputObject` which already takes a string and converts to Guid? Or am I missing something here?
> 
> The cmdlet hasn't `InputObject` parameter. The parameter is mainly for pipeline scenario. In the case it is better to have `ConvertTo-Guid` cmdlet.

@iSazonov `New-Guid` for 7.5 preview release does have `-InputObject` as parameter: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/new-guid?view=powershell-7.5#-inputobject

This was added in #20014.

**Comment by iSazonov** [comment]
Ah, it escaped my attention. Then it is not clear why an additional parameter is needed at all.

**Comment by Meir017** [comment]
this should include the new Guid v7 creation API - (issue https://github.com/dotnet/runtime/issues/103658, pr https://github.com/dotnet/runtime/pull/104124)

**Comment by amorenojr** [comment]
I would also like to see the ability to generate a unique guid based on a list of existing guids. Example code below.

```powershell
function Get-Guid {
    [CmdletBinding()]
    param (
        [Parameter(Mandatory=$false)]
        [Alias("g")]
        [string[]]
        $ExistingGuids
    )

    $guid = New-Guid

    if ($ExistingGuids.Count -gt 0) {
        while ($ExistingGuids -contains $guid) {
            $guid = New-Guid
        }
    }

    return $guid
}

```

**Comment by TobiasPSP** [comment]
The purpose of "Guids" is to guarantee unique numbers with no need of central issuer. That's coincidentally what the "GU" stands for in their name. 

A 6-out-of-49 lottery is 2^24, the GUID range is 2^128 I believe. 

You'd rather win the lottery jackpot five times in a row than accidentally drawing the same GUID twice. IT is working successfully on this promise for a good decade.

If you do need a blacklist to double-check for duplicates, I'd keep this outside the cmdlet code. ;-)

**Comment by amorenojr** [comment]
> The purpose of "Guids" is to guarantee unique numbers with no need of central issuer. That's what the "GU" should stand for in their name. 
> 
> A 6-out-of-49 lottery is 2^24, the GUID range is 2^128 I believe. 
> 
> You'd rather win the lottery jackpot five times in a row than accidentally drawing the same GUID twice. It is not *impossible* to draw the same GUID twice. It is just kinda unlikely. IT is working successfully on this assumption for a good decade.
> 
> If you do need a blacklist to double-check for duplicates, I'd keep this outside the cmdlet code. ;-)
> 
> 

I agree. However, I hit the lottery on this once and spent weeks trying to figure it out. Now I just want to be cautious. 

If it doesn't belong in the cmdlet, I will just maintain it as a stand-alone feature I use.

Unless anyone else wants to chime in?

**Comment by TobiasPSP** [comment]
>I hit the lottery on this once and spent weeks trying to figure it out. Now I just want to be cautious.

I hear you ;-)

Imho, New-Guid should be focused: create a new guid. Any post-processing such as sorting out unwanted ones should be user code.

A reason for involuntary guid duplicates btw can be that a person created a new guid by manually flipping a few numbers from an existing one. Whenever an org starts to accept guids that are biased, duplicates are likely to occur.
